cirq-core/cirq/ops/gateset.py

```diff
@@ -0,0 +1,292 @@
+        self._gates = frozenset(
+            [g if isinstance(g, GateFamily) else GateFamily(gate=g) for g in gates]
+        )
+
+    @property
+    def name(self) -> Optional[str]:
+        return self._name
+
+    @property
+    def gates(self) -> FrozenSet[GateFamily]:
+        return self._gates
+
+    def __contains__(self, item: Union[raw_types.Gate, raw_types.Operation]) -> bool:
+        """Check for containment of `item` in any of the member `GateFamily`.
+
+        This method just forwards containment checks to the underlying
+        `GateFamily` objects. It returns `True` if any of the member gate
+        families can accept `item`.
+
+        To validate whether a `cirq.CircuitOperation`/`cirq.OP_TREE`/
+        `cirq.Circuit` is made up entirely of gates from this Gateset, use the
+        `validate` method given below.
+
+        Args:
+            item: The `cirq.Gate` or `cirq.Operation` instance to check
+                containment for.
+        """
+        for gate_family in self.gates:
+            if item in gate_family:
+                return True
+        return False
+
+    def validate(
+        self,
+        circuit_or_optree: Union['cirq.AbstractCircuit', op_tree.OP_TREE],
+        *,
+        unroll_circuit_op: bool = True,
+        accept_global_phase: bool = True,
+    ) -> bool:
+        """Validates gates forming `circuit_or_optree` should be contained in Gateset.
+
+        Args:
+            circuit_or_optree: The `cirq.Circuit` or `cirq.OP_TREE` to validate.
+            unroll_circuit_op: If True, `cirq.CircuitOperation` is recursively
+                validated by validating the underlying `cirq.Circuit`.
+            accept_global_phase: If True, `cirq.GlobalPhaseOperation` is accepted.
+        """
+        # To avoid circular import.
+        from cirq.circuits import circuit
+
+        optree = circuit_or_optree
+        if isinstance(circuit_or_optree, circuit.AbstractCircuit):
+            optree = circuit_or_optree.all_operations()
+        return all(
+            self._validate_operation(
+                op, unroll_circuit_op=unroll_circuit_op, accept_global_phase=accept_global_phase
+            )
+            for op in op_tree.flatten_to_ops(optree)
+        )
+
+    def _validate_operation(
+        self,
+        op: raw_types.Operation,
+        *,
+        unroll_circuit_op: bool = True,
+        accept_global_phase: bool = True,
+    ) -> bool:
+        # To avoid circular import.
+        from cirq.circuits import circuit_operation
+
+        if isinstance(op, raw_types.TaggedOperation):
+            return self._validate_operation(
+                op.sub_operation,
+                unroll_circuit_op=unroll_circuit_op,
+                accept_global_phase=accept_global_phase,
+            )
+        elif isinstance(op, circuit_operation.CircuitOperation) and unroll_circuit_op:
+            return self.validate(
+                op.mapped_circuit(deep=True),
```

Comment:
Yes, this will be unroll the circuit `X` times. The reason I used this is because we need to recursively flatten any `CircuitOperation`s present inside the circuit contained by top level `CircuitOperation`. I'm not sure if there's a way to just get a list of all unique operations inside a `CircuitOperation` and avoid any repetitions. 

@95-martin-orion Does such a functionality already exist? If not, Do you think this would be a useful function to add to the CircuitOperation class or should I write a custom unroller here?

Reply:
`op.circuit` contains all unique operations in the `CircuitOperation` modulo measurement key changes. You'll need to manually apply the parameter mapping, though - that can be done with:
```
protocols.resolve_parameters(op.circuit, op.param_resolver, recursive=False)
```
Parameter values are the same for every iteration. We originally considered providing an "iterator" parameter which changed for each repetition, but that feature didn't make it into the released version.

Reply:
I think it would be nice if we can add a method on `CircuitOperation` which allows us to iterate over / returns us a list of all mapped operations without the repetition. 

For eg, even if I apply the `resolve_parameters` mentioned above, the qubit mapping would also need to applied manually at the call site -- and this can affect whether an operation is "valid" or not for topology aware gate families. 

In general, applying CircuitOperation specific mappings outside of the class sounds like a bad idea because if any new kind of mapping is introduced in the future, the call sites would also need to be changes. Thoughts?

Reply:
I've made the change here to explicitly use the underlying `op.circuit` and resolve parameters and apply qubit mappings, but I still think having a utility method on `CircuitOperation` would be the right thing to do here.